### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by cccelik**

### CODE REVIEW

File: src/MolSim.h
Comment (SamNewcome) [inline src/MolSim.h:1]: What was the reason to create a separate header?

File: src/MolSim.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                ("h,help", "print help message")
+                ("i,input", "input file path", cxxopts::value<std::string>())
+                ("d,deltaT", "Set deltaT", cxxopts::value<double>()->default_value("0.014"))
+                ("e,endTime", "Set endTime", cxxopts::value<double>()->default_value("1000"))
+                ("o,output", "Set Outputwriter (VTK or XYZ)", cxxopts::value<std::string>())
+                ("c,calculator", "Set Calculator", cxxopts::value<std::string>())
+                ("l,logLevel", "Set log level", cxxopts::value<std::string>());
+
+
+        auto res = options.parse(argc, argv);
```
Comment (SamNewcome) [inline src/MolSim.h:87]: `res` is not a descriptive name

File: src/Objects/Cuboid.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+//
+
+#ifndef CUBOID_H
+#define CUBOID_H
+#include <array>
+
+#include "ParticleContainer.h"
+
+
+class Cuboid {
```
Comment (SamNewcome) [inline src/Objects/Cuboid.h:12]: Classes themselves should be documented

File: src/Objects/Particle.cpp [hunk 3]
```diff
@@ -4,17 +4,20 @@
...
  *      Author: eckhardw
  */
-
+#include "spdlog/spdlog.h"
 #include "Particle.h"
 
 #include <iostream>
 #include "utils/ArrayUtils.h"
 
+int ids = 0;
```
Comment (SamNewcome) [inline src/Objects/Particle.cpp:13]: Be super careful about this! I would generally recommend not doing this because it gives you no real control over what IDs the particles have. This could result in duplicate particle IDs or particle IDs changing throughout a simulation or series of simulations when combined with advanced features like checkpointing or distributed memory parallelisation.

File: src/Objects/ParticleContainer.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    return p;
+                }
+            }
+        } else { //jump to pos id and compare, move to left if ids don't match (particle with smaller id has been deleted)
+            int i = id;
+            while (particles.at(i).getID() != id) {
+                i--;
+            }
+            return particles.at(i);
+        }
```
Comment (SamNewcome) [inline src/Objects/ParticleContainer.cpp:58]: I think this is quite messy, error-prone, and gives no performance gain. Correct me if I've misunderstood:

- The `if`-clause only triggers if a particle has been deleted and the `particles` vector is resized to account for this. Here a simple linear search is performed.
- If `id + 1 > particles.size()`, the `else`-clause is taken.
- Here your particle `.at(i)`'s ID is tested. Except, this particle will be beyond the length of the vector, so this would easily lead to a segmentation fault. I suspect this problem was never found as I expect this `else`-clause never gets triggered.
- Even if this did work, you are still just performing a linear search, but in reverse, so there is no performance gain.

I suspect this is why your tests fail when run with an address sanitizer.

If I am right, please remove this and just use lines 60-64 or my suggestion below. If not, please let me know.

A lot of good "standard" algorithms are implemented in `<algorithm>`. This includes `std::find`, which does what you want.

File: src/Objects/ParticleGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                // calculate x,y,z coordinates
+                double x_coor = x[0] + i*h;
+                double y_coor = x[1] + j*h;
+                double z_coor = x[2] + k*h;
+                std::array<double, 3> particle_pos = {x_coor,y_coor,z_coor};
+                // get the maxwell velocity
+                std::array<double, 3> maxwell_vel = maxwellBoltzmannDistributedVelocity(mv,2);
+                // get the initial velocity
+                std::array<double, 3> vel = v;
+                // add maxwell velocity to initial velocity? --> I don't know if this is what we are supposed
```
Comment (SamNewcome) [inline src/Objects/ParticleGenerator.cpp:32]: Yes this is fine

File: src/Calculator/Calculator.h
Comment (SamNewcome) [inline src/Calculator/Calculator.h:1]: I think it is misleading that your base class implements gravity force, which is then overridden by, e.g. the Lennard-Jones force calculation. I would have `Calculator` simply as a base class which "`GravityCalculator`" and `LennardJonesCalculator` can inherit from.

File: src/Calculator/Calculator.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            }
+            for (auto &pair : particleContainer.getParticlePairs()) {
+                Particle &p1 = pair.first.get();
+                Particle &p2 = pair.second.get();
+                std::array<double, 3> sub = operator-(p2.getX(), p1.getX());
+                double normCubed = ArrayUtils::L2Norm(sub);
+                //prevent division by 0
+                if (normCubed < 1e-8) {
+                    continue;
+                }
```
Comment (SamNewcome) [inline src/Calculator/Calculator.h:50]: This shouldn't ever occur. If it does, it means that something is wrong in your simulation e.g. you have duplicate particles or your step size was chosen too high. You could put a warning here.

File: src/Objects/ParticleContainer.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                }
+            }
+        
+        SPDLOG_WARN("Particle not found");
+        throw std::runtime_error("Particle not found");
+    }
+
+    Particle& ParticleContainer::getParticle(Particle& p) {
+        return ParticleContainer::getParticle(p.getID());
+    }
```
Comment (SamNewcome) [inline src/Objects/ParticleContainer.cpp:72]: I don't get the point of this function: you are passing in the particle which you want to get.

File: tests/Calculator/CalculatorTest.cpp [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    test.addParticle(k);
+
+    test.initializePairsVector();
+
+
+    Calculators::Calculator calc;
+    calc.calculateF(test); 
+
+    for (Particle& p: test) {
+        ASSERT_EQ(old, test.getParticle(p.getID()).getOldF());
```
Comment (SamNewcome) [inline tests/Calculator/CalculatorTest.cpp:55]: Why use `ASSERT` here? I recommend using `EXPECT` unless there is a reason to abort the test. See https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

File: tests/Calculator/CalculatorTest.cpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        EXPECT_NEAR(0.075, t.getX()[0], 0.00001);
+        EXPECT_NEAR(0.075, t.getX()[1], 0.00001);
+        EXPECT_NEAR(0.075, t.getX()[2], 0.00001);
+    }
+
+}
+
+/*Checks that calculateF() correctly updates the force between particles*/
+TEST(CalculatorTest, correctFIJcalculations) {
+    ParticleContainer test;
```
Comment (SamNewcome) [inline tests/Calculator/CalculatorTest.cpp:33]: `test` is not a very clear name. I'd suggest `testContainer`

**Review by SamNewcome — APPROVED** [review]
Note
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._
----------------------
Quite nice, especially a lot of the refactoring and the unit tests.

The main problem is that you don't run your tests in your CI and that your particle getter is memory unsafe. If your CI ran the tests, you probably would've spotted this. I discuss the getter logic in more detail with the code.

My general advice: balance performance optimisations with clarity of code. Clearer code is easier to maintain and easier to spot bugs with. The getter is not used in performance critical sections, so clarity is more important. Even if it was used in a performance critical section, only merge the optimised version over the simple version if you have hard evidence that it is more performant.

P.s., the CI wasn't included in the PR. The issue there is that you don't turn on the CMake flag for tests.

Please fix the CI running of tests and the particle getter. Everything else is minor.

File: src/MolSim.h
Comment (cccelik) [reply src/MolSim.h:1]: MolSim :cpp

In Assignment 1 on Moodle for the review, it was stated that we had too much stuff in the molsim.cpp class so we decided to move the actual parsing to a header file to declutter our main function and class. We wanted to make the execution and the main function more user friendly and easier to read.

File: src/Calculator/Calculator.h
Comment (cccelik) [reply src/Calculator/Calculator.h:1]: While implementing, we thought that we could have the basic calculations in the header file since we only override the calculations between the particle pairs. Since the whole force calculation is not overridden, we thought creating a new class for Gravity Calculator was not useful at the moment, but we can add a new class in the assignment 3 branch if necessary.

File: src/Calculator/Calculator.h (hunk 6 above)
Comment (cccelik) [reply src/Calculator/Calculator.h:50]: We added this as an edge case to make sure nothing goes wrong. Initially in assignment 1 we were checking for == 0 to prevent division by zero but Mr. Mishra has suggested we change the check as it is right now:


edge case

**Comment by cccelik** [comment]
> P.s., the CI wasn't included in the PR. The issue there is that you don't turn on the CMake flag for tests.

We were not sure if we were supposed to push the CI to the new branch or the master branch since we weren’t sure what the proper way to make GitHub actions work was, so we decided to add it directly into the master branch. In our CMakeList file, the tests are on by default unless they are specifically turned off. If you checked the ci pipeline by using the actions tab, you might have looked at the recent executions. We were trying to adapt it for assignment 3, and currently, we are facing an issue where the pipeline does not execute at all. But since it is merged from the initially forked repo into our master branch, the tests are not implemented thus, there are no tests to execute in the top actions under the tab. If you check the checks executed under “Assignment2” under the actions tab from a week ago, it shows all the tests we have written have been executed and also passed for our code. This is the latest execution of the pipeline before the deadline for assignment 2. I will also be attaching a picture of this. Unless there is another mistake we cannot see, we don’t think there is an issue with the pipeline for assignment 2. We also did not get a warning about a memory issue during the execution of the tests neither on GitHub nor locally.


Tests in pipeline

pipeline 2

File: src/Objects/ParticleContainer.cpp (hunk 4 above)
Comment (StefXQB) [reply src/Objects/ParticleContainer.cpp:58]: This method operates under the assumption that few particles have been deleted, that the particle is indeed contained in the ParticleContainer and that the particles were added to the container in order of ascending ids, ideally starting with the id 0. Let us look at a Particle with id x. If no particles with an id smaller than x have been deleted and the ids start with 0, the Particle with id x is located at position x. If a few particles with an id < x have been deleted, we only need to move forward a few steps from position x in order to find the correct particle. Therefore, we can improve performance under certain circumstances, which are usually given during normal execution. The algorithm described here is performed in the **else-clause**.

If too many particles have been deleted (id+1 > particles.size()), the **if-clause** is taken to prevent accessing out-of-bounds when trying to access the particle at position id because too many particles have been deleted.

In the **else-clause**, no out-of-bounds memory access can occur if the Particle is really contained in the ParticleContainer. Even if the ids do not start with 0, as long as the particles are added in order of ascending id, the Particle with id x is guaranteed to be at a position <= x. This is due to the fact that new particles always get added to the Container via pushback, which adds them at the very last position of the vector, and the fact that the smallest possible id for a particle to have is 0.

While the method is not inherently wrong and does what it should when used correctly (we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline) and can improve performance under the right circumstances, I agree that the method is convoluted and confusing. I will be heeding your advice and following your suggestions to replace this method to protect from errors caused by incorrect usage and improve code clarity. We are currently only using this method for tests, therefore performance is not an issue.

File: src/Objects/ParticleContainer.cpp (hunk 7 above)
Comment (StefXQB) [reply src/Objects/ParticleContainer.cpp:72]: This function was added to circumvent a problem that we encountered when writing tests: If we create a Particle, add it to a ParticleContainer and then perform an operation on the ParticleContainer (e.g., updateX), we noticed that if we get the values from the particle we will find that they are unchanged, even when using references. We are unsure as to why this happens, but we found that we can access the new values of particle attributes by getting the particles from the ParticleContainer, either via iterator or via a getParticle function when trying to access a specific Particle.

Comment (SamNewcome) [reply src/Objects/ParticleContainer.cpp:72]: I'm guessing that you create a copy of the particle when you add it to the container. Don't quote me on that however - maybe check using the log messages.

File: src/MolSim.h
Comment (SamNewcome) [reply src/MolSim.h:1]: These file types are really there for separate reasons: the header files allow you to reference declarations of functions/classes/stuff in other files without having to import the implementation which is kept in the `.cpp`. This keeps from project smaller once built and improves compilation time. For now compilation time is not an issue, but with bad practice and a larger project, this becomes a real problem.

As such, you are just moving implementation clutter from the `.cpp` to the `.h` which is even less suitable.

The better solution is to use entirely separate files for a lot of this functionality and keep the main file just as a cpp. It is fine is your main file really just parses the input (potentially using a separate class with separate files) and then uses a separate class (with separate files) for the actual simulation. This is cleaner and may help with compilation times.

File: src/Objects/ParticleContainer.cpp (hunk 4 above)
Comment (SamNewcome) [reply src/Objects/ParticleContainer.cpp:58]: > This method operates under the assumption that few particles have been deleted, that the particle is indeed contained in the ParticleContainer and that the particles were added to the container in order of ascending ids, ideally starting with the id 0. Let us look at a Particle with id x. If no particles with an id smaller than x have been deleted and the ids start with 0, the Particle with id x is located at position x. 

Makes sense.

> If a few particles with an id < x have been deleted, we only need to move forward a few steps from position x in order to find the correct particle. 
> Therefore, we can improve performance under certain circumstances, which are usually given during normal execution. 

I guess the idea is that it is faster to start iteration from the back fo the vector backwards than from the start of the vector forwards. I don't see this as faster (in general - of course, if the particle you are looking for is closer to the back, then this is better - but you don't know this).


> The algorithm described here is performed in the **else-clause**.
> 
> If too many particles have been deleted (id+1 > particles.size()), the **if-clause** is taken to prevent accessing out-of-bounds when trying to access the particle at position id because too many particles have been deleted.

Correct.

> 
> In the **else-clause**, no out-of-bounds memory access can occur if the Particle is really contained in the ParticleContainer. Even if the ids do not start with 0, as long as the particles are added in order of ascending id, the Particle with id x is guaranteed to be at a position <= x. This is due to the fact that new particles always get added to the Container via pushback, which adds them at the very last position of the vector, and the fact that the smallest possible id for a particle to have is 0.

No this is not correct. Suppose you add 10 particles to the container with ascending order. Then you delete particles 4-6. Then you try to look for particle with ID 9.

So you check `particles.at(9).getID() != id`.

If you have already resized `particles`, such that it is size 7 (no empty spaces), then this is out-of-bounds.

I guess the case where I do see some advantage is if you are specifically only/mostly deleting particles with lower IDs and not resizing the vector. But here, you will just be growing the vector larger and larger, which also is not suitable or performant. 

Better solutions to this problem involve not growing the vector more than needed. You can do swap deletes (swap to-be-deleted particle with end particles, then pop that off the vector) or you can just periodically resort the vector to remove empty spaces in the vector. 

> 
> While the method is not inherently wrong and does what it should when used correctly (we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline) and can improve performance under the right circumstances, I agree that the method is convoluted and confusing. I will be heeding your advice and following your suggestions to replace this method to protect from errors caused by incorrect usage and improve code clarity. We are currently only using this method for tests, therefore performance is not an issue.

> we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline

Do you mean when running the tests or just the main program?

File: src/Calculator/Calculator.h (hunk 6 above)
Comment (SamNewcome) [reply src/Calculator/Calculator.h:50]: So Manish's suggestion is definitely better as it deals with floating point errors. But I would suggest not having either.

**Comment by SamNewcome** [comment]
I see your point about the CI. Please don't merge these changes separately, it works fine if you update this within the branch you are working on. But I accept that you are typically running the ctests in the CI.

This will likely have an impact on the grade _but_ I still get failures when I run your tests with an address sanitiser and I want to check for why there is a discrepancy between what you are doing and what I am doing. I don't have the time at the moment to investigate this. Once I've done so, I will rediscuss your grade with Manish.

File: src/Objects/ParticleContainer.cpp (hunk 4 above)
Comment (StefXQB) [reply src/Objects/ParticleContainer.cpp:58]: > No this is not correct. Suppose you add 10 particles to the container with ascending order. Then you delete particles 4-6. Then you try to look for particle with ID 9.
> 
> So you check particles.at(9).getID() != id.
> 
> If you have already resized particles, such that it is size 7 (no empty spaces), then this is out-of-bounds.

If we are looking for a particle with id 9 and size is 7, then `(int)particles.size() < id + 1` evaluates to true, so we are taking the **if-clause instead of the else-clause**. We are performing linear search and therefore not accessing out-of-bounds. This if-clause is there specifically to prevent access out-of-bounds at the back of the vector in the else-clause. Access out-of-bounds at the front of the vector is prevented by the fact that the particle will definitely be found within the first x entries if it is present within the vector, as I described earlier.
To prove that your example does not cause an error, I have written a new test case for it, which I can push to the Assignment 2 branch if you wish to try it yourself.
```
TEST(ParticleContainerTest, testGetParticleExample) {
    ParticleContainer testContainer;
   
    for (int i= 0; i < 4; i++) {
        Particle p(0); //0 is the type, not the id
        testContainer.addParticle(p);
}
    Particle p4(0); //id = 4
    testContainer.addParticle(p4);
    Particle p5(0); //id = 5
    testContainer.addParticle(p5);
    Particle p6(0); //id = 6
    testContainer.addParticle(p6);

    for (int i= 7; i < 10; i++) {
        Particle p(0); //0 is the type, not the id
        testContainer.addParticle(p);
    }
    EXPECT_EQ(testContainer.getParticles().size(), 10);

    for (Particle p : testContainer) {
        std::cout << p.getID() << ", ";
    }
    std::cout << std::endl;

    testContainer.removeParticle(p4);
    testContainer.removeParticle(p5);
    testContainer.removeParticle(p6);

    EXPECT_EQ(testContainer.getParticles().size(), 7);

    for (Particle p : testContainer) {
        std::cout << p.getID() << ", ";
    }
    std::cout << std::endl;

    EXPECT_EQ(testContainer.getParticle(9).getID(), 9);
}

```
Running this single test with an address sanitizer, I get the following output:

```
stef@laptop:~/molsim/oldVersion/MolSim/build$ ./Test --gtest_filter=ParticleContainerTest.testGetParticleExample
Note: Google Test filter = ParticleContainerTest.testGetParticleExample
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ParticleContainerTest
[ RUN      ] ParticleContainerTest.testGetParticleExample
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 
0, 1, 2, 3, 7, 8, 9, 
[       OK ] ParticleContainerTest.testGetParticleExample (0 ms)
[----------] 1 test from ParticleContainerTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.

```
It is necessary to make sure that only this single test is run to guarantee that the IDs start with 0. If other test cases are run as well, it is possible that other Particles are generated first and therefore the IDs may start with a higher number, which leads to a test failure because the particle with id 9 cannot be found. During normal non-parallelized execution of a simulation, the particles always start with id 0 because only one ParticleContainer is generated. If this changes in the future, the particle id initialization and the getParticle function must be adapted accordingly.

> I guess the case where I do see some advantage is if you are specifically only/mostly deleting particles with lower IDs and not resizing the vector. But here, you will just be growing the vector larger and larger, which also is not suitable or performant.
> Better solutions to this problem involve not growing the vector more than needed. You can do swap deletes (swap to-be-deleted particle with end particles, then pop that off the vector) or you can just periodically resort the vector to remove empty spaces in the vector.

We are using std::vector::erase to remove particles from the container, so the vector does not grow more than needed. When removing a particle, this function resizes the vector and no empty spaces are left within the vector. 

>  Do you mean when running the tests or just the main program?

Both don’t produce any error with an address sanitizer, neither locally nor in the pipeline. We suspect that you may be including the test addParticleToPairs, which currently fails the address sanitizer both locally and in our ci pipeline. However, this test does not belong to Assignment 2 branch, but is instead part of the Assignment 3 branch and tests new functions added as part of Assignment 3 and therefore should not influence the grading for Assignment 2. Furthermore, the error is caused not by getParticle(), but instead by accessing deleted particles. I have temporarily removed this test, please try running the tests again or better yet, make sure to use the last commit to Assignment 2 before the deadline (the commit with the message “changed log messages from info to debug, parse -l first” to run the tests. This is also the commit that shows all 11 tests running in the CI pipeline without issue.

Comment (SamNewcome) [reply src/Objects/ParticleContainer.cpp:58]: I am building with gcc 11.4.0, and the commit is 09b117a7ba7f7347b1b40d2d497737d12b87c000 . Inside a build folder, I am running `cmake -DCMAKE_CXX_FLAGS="-fsanitize=address" ..` and then `make`. I then run `ctest`.

The test failures are non-deterministic. I ran `ctest` three times consecutively:

The first time, I get
```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............   Passed    0.02 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............   Passed    0.01 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............   Passed    0.01 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......   Passed    0.01 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.01 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.02 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........***Exception: SegFault  0.10 sec

91% tests passed, 1 tests failed out of 11

Total Test time (real) =   0.33 sec

The following tests FAILED:
	 11 - ParticleGeneratorTest.checkOtherVariables (SEGFAULT)
```

The second time, I get
```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............***Exception: SegFault  0.09 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............***Exception: SegFault  0.09 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............***Exception: SegFault  0.09 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......***Exception: SegFault  0.09 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.02 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.01 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........   Passed    0.01 sec

64% tests passed, 4 tests failed out of 11

Total Test time (real) =   0.54 sec

The following tests FAILED:
	  1 - CalculatorTest.correctXcalculations (SEGFAULT)
	  3 - CalculatorTest.correctVcalculations (SEGFAULT)
	  6 - CuboidTest.testGettersAndConstructor (SEGFAULT)
	  8 - ParticleContainerTest.PairsAfterAddParticle (SEGFAULT)
```

The third time I get

```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............   Passed    0.01 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............   Passed    0.02 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............***Exception: SegFault  0.10 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......   Passed    0.01 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.01 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.01 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........***Exception: SegFault  0.11 sec

82% tests passed, 2 tests failed out of 11

Total Test time (real) =   0.38 sec

The following tests FAILED:
	  6 - CuboidTest.testGettersAndConstructor (SEGFAULT)
	 11 - ParticleGeneratorTest.checkOtherVariables (SEGFAULT)
```

With a verbose output, I get `AddressSanitizer:DEADLYSIGNAL`.

With GCC 14.2.1 on my laptop, I don't get any errors. I'm not sure why there is a discrepancy between the CI and my workstation.

I understand this is tricky, if this is only causing a problem on my workstation. We won't penalise you for a problem that is only producible on my local machine and not on the CI. Please avoid changing the CI in the master to make it easier for us to spot this.

You are correct regarding the `if`/`else` clauses; that was my misunderstanding!

> We are using std::vector::erase to remove particles from the container, so the vector does not grow more than needed. When removing a particle, this function resizes the vector and no empty spaces are left within the vector.

Ah okay. Yes, this is correct. But in this case, I really don't see the point of this reverse linear search.

**Comment by SamNewcome** [comment]
I've updated your mark.

After the worksheet 3 submission, please do not merge anything into the master branch anymore (including CI) outside of the Assignment hand-in branches. Please also do not merge any changes from the `master` into any Assignment hand-in branch after the submission.
